### ISSUE TITLE
Poor inference when passing ref nullable variable that happens to be non-null

### ISSUE THREAD

**Issue by gafter** [dotnet/roslyn#34717]
The type parameter below is inferred as `string` rather than `string?`, even though the first parameter is a ref to a variable declared `string?` and the second is a null literal.

``` c#
#nullable enable
using System.Collections.Generic;
using System.Threading;

class Program
{
    static void Main() { }

    static string? _field;
    static void Foo(IEnumerable<object> source)
    {
        if (_field != null)
        {
            string? s = Interlocked.Exchange(ref _field, null); // unexpected warning
        }
    }
}
```

Can we do better?

Originally observed at https://github.com/dotnet/coreclr/pull/23609#discussion_r270700806

**Comment by jcouv** [comment]
It looks like this no longer produces a warning ([sharplab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQBMQGoA+BiAdgVwBtCJhC4ACOXU8gWACgABABgqYEYAWAbkdfYcAdABUAFgjgQUAS1wBzPg34AmQQHZGAb0YU9ggGzsuFALIQ5ACgCUFLRQC+jXfs5HOLAPwUA+gDMZOEIUJX1DYwoAMQB7aMtOAGYAHmjgACs4AGMYAD4KKGj8BEy4axc9HQYwsJk/Ckt/QOCKAEIAXgoCYjKq6oryvtcOL3yKDoBJXHgEQmjMgGs4FCEAUQAPTLEIBThLSTrGoJQkTqJCax4KAHorinxcODWAByz4FAoAdwgEXDl5AbCTl6eiBDiAA===))
